### DISCUSSION TITLE
Electricity-map Plugin

### DISCUSSION THREAD

**Discussion by yuu1127** [Green-Software-Foundation/carbon-aware-sdk#67] [General]
Hi, after I read the code for the 0.1 previews on Github, 
it seems possible to adopt a similar style for the electricity map as same as Watttime plugin
since the electricity map API for the personal version also provides the latest carbon intensity.
(https://electricitymap.org/)
For example usage, 
```
**API call by latitude, longitude**
curl 'https://api.co2signal.com/v1/latest?lon=6.8770394&lat=45.9162776'
  -H 'auth-token: myapitoken'
->
{
    "countryCode": "FR",
    "data": {
        "carbonIntensity": 92.97078744790771,
        "datetime": "2017-02-09T08:30:00.000Z",
        "fossilFuelPercentage": 12.028887656434616
    },
    "status": "ok",
    "units": {
        "carbonIntensity": "gCO2eq/kWh"
    }
}

**API call by countryCode**
curl 'https://api.co2signal.com/v1/latest?countryCode=JP-TK'
  -H 'auth-token: myapitoken'
->
{
    "countryCode": "JP-TK",
    "data": {
        "carbonIntensity":499 ,
        "datetime": "2022-06-29T00:00:00.000Z",
        "fossilFuelPercentage": 74.49
    },
    "status": "ok",
    "units": {
        "carbonIntensity": "gCO2eq/kWh"
    }
}
```
Only a big difference is "areas" which they support `BA - Balancing Authority for Watttiem`  `zone - country-code for electricity map` and electricity map doesn't provide forecast and historical data for the personal version.
Electricity map supports **more** countries' data including many Asia such as India, China and others which Watttime doesn't support. 
For these reasons, I'm thinking to make a new branch called "electricity-map" and prepare development to catch up with the Wattitme plug-in release.
What do you guys think? Any comments are welcome.

**Comment by luwangshell** [comment]
also, electricity-map support the breakdown on the carbon emission, which can be very useful.

**Reply by luwangshell** [reply]
one example:

{
  "zone": "DK-DK1",
  "datetime": "2020-01-04T00:00:00.000Z",
  "updatedAt": "2022-04-07T09:35:03.914Z",
  "createdAt": "2022-02-04T15:49:58.284Z",
  "powerConsumptionBreakdown": {
    "nuclear": 65,
    "geothermal": 0,
    "biomass": 188,
    "coal": 195,
    "wind": 1642,
    "solar": 0,
    "hydro": 13,
    "gas": 72,
    "oil": 8,
    "unknown": 5,
    "hydro discharge": 0,
    "battery discharge": 0
  },

**Reply by yuu1127** [reply]
Yes, this feature would be also useful ( Unfortunately this feature seems supported by commercial ones, not personal ).

**Comment by vamsi12a2** [comment]
Hi, I just want to highlight a minor limitation of this personal version of API. It has a rate limit of 30 requests per hour and a max 1 request per second. The same will propagate to our SDK when we use it. Are we planning to handle this one?

**Reply by yuu1127** [reply]
Thanks @vamsi12a2,  I also noticed that if keep sending API requests, it shows an error message "API rate limit exceeded",
We need to consider this for future usage.  (e.g. save Cache data that other user requested somewhere
or think about using the Commercial version )

**Comment by yuu1127** [comment]
I made an issue here #71.

As far as I touched electricityMap Personal version.
The big difference from Watttime is below

1. can return gCO2eq/kWh directly even if inputs are latitude and longitude (for Watttime return nearby Balancing Authority Area)
2. acquiring authorization token only by e-mail (Watttime has API feature for this so can send request and get token inside code )
3. the components of json response are a bit different (for electriciyMap nested property)

e.g.
for Watttime
`{"point_time": "2022-05-02T00:15:00.000Z", "value": 38.0, "frequency": 300, "market": "RTM", "ba": "CAISO_NORTH", "datatype": "MOER", "version": "3.0"}`
for electricityMap
`{"_disclaimer":"This data is the exclusive property of electricityMap and/or related parties. If you're in doubt about your rights to use this data, please contact api@co2signal.com","status":"ok","countryCode":"FR","data":{"datetime":"2022-06-27T05:00:00.000Z","carbonIntensity":72,"fossilFuelPercentage":10.92},"units":{"carbonIntensity":"gCO2eq/kWh"}}`
